File: modules/admin/impl/default-ts/middleware.ts
import type { Context, Next } from 'hono'
import type { IAuthService } from '../../../../contracts/ts/auth'

/**
 * adminRequired is Hono middleware that ensures the request is from an admin user.
 * Must be used AFTER auth middleware that sets the session token.
 */
export function adminRequired(auth: IAuthService) {
  return async (c: Context, next: Next) => {
    const token = c.req.header('Authorization')?.replace('Bearer ', '')
    if (!token) {
      return c.json({ error: 'Unauthorized' }, 401)
    }

    const user = await auth.validateToken(token)
    if (user.role !== 'admin') {
      return c.json({ error: 'Forbidden: admin access required' }, 403)
    }

    c.set('adminUser', user)
    await next()
  }
}

File: modules/admin/impl/default-go/middleware.go
package admin

import (
	"net/http"
	"strings"

	contracts "github.com/dat2503/modkit/contracts/go"
)

// AdminRequired is HTTP middleware that ensures the request is from an admin user.
// Must be used AFTER auth middleware that sets the Authorization header.
func AdminRequired(auth contracts.AuthService) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			token := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
			if token == "" {
				http.Error(w, `{"error":"Unauthorized"}`, http.StatusUnauthorized)
				return
			}

			user, err := auth.ValidateToken(r.Context(), token)
			if err != nil {
				http.Error(w, `{"error":"Unauthorized"}`, http.StatusUnauthorized)
				return
			}

			if user.Role != "admin" {
				http.Error(w, `{"error":"Forbidden: admin access required"}`, http.StatusForbidden)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}
